File: PALIN.py
"""
---------------luyencode.net----------------------------------
---------------Problem: PALIN-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-06 23:43:28.020814+07:00---------
"""
in1 = lambda: int(input())
in2 = lambda: list(map(int, input().split()))
MOD = 1000000007
# --------------- Solve ---------------  # 
def preprocess(s):
    """Add special characters to handle even-length palindromes."""
    return '#' + '#'.join(s) + '#'

def manacher(s):
    """Manacher's algorithm to find the longest palindromic substring."""
    T = preprocess(s)  # Transform the string
    n = len(T)
    P = [0] * n  # Array to hold the length of palindromes
    C = R = 0  # Center and right boundary

    for i in range(1, n - 1):
        mirror = 2 * C - i  # Mirror of i around center C

        if R > i:
            P[i] = min(R - i, P[mirror])  # Use the mirror value

        # Attempt to expand the palindrome centered at i
        while i + P[i] + 1 < n and i - P[i] - 1 >= 0 and T[i + P[i] + 1] == T[i - P[i] - 1]:
            P[i] += 1

        # Update the center and right boundary if needed
        if i + P[i] > R:
            C = i
            R = i + P[i]

    # Find the maximum length and its center
    max_length = 0
    center_index = 0
    for i in range(1, n - 1):
        if P[i] > max_length:
            max_length = P[i]
            center_index = i

    # Recover the longest palindromic substring
    start = (center_index - max_length) // 2  # Calculate start index in original string
    return s[start:start + max_length]

s = input()

print(manacher(s))
##########################################

File: PALIN.cpp
/*
---------------luyencode.net----------------------------------
---------------Problem: PALIN-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-07 00:08:20.262554+07:00---------
*/
#include <iostream>
#include <iomanip>
#include <vector>
#include <queue>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <assert.h>
#include <algorithm>
#include <functional>
#include <numeric>
#include <math.h>

#define FAST_IO ios::sync_with_stdio(false);cin.tie(0);cout.tie(0);
using namespace std;

const int MOD = 1000000007;

int main() {
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    FAST_IO;
    // --------------- Solve -------------------
    string s;
    cin >> s;
    int ans = 0;
    int l = s.length();

    vector<vector<int>> dp(l, vector<int>(l, 0));

    for (int i = 0; i < l; ++i) {
        dp[i][i] = 1;
    }

    for (int i = 0; i < l - 1; ++i) {
        if (s[i] == s[i + 1]) dp[i][i + 1] = 2;
        else dp[i][i + 1] = 1;
    }

    for (int i = 2; i < l; ++i) {
        for (int j = 0; j < l - i; ++j) {
            if (s[j] == s[i + j]) {
                dp[j][i + j] = 2 + dp[j + 1][i + j - 1];
            }
            else {
                dp[j][i + j] = max(dp[j + 1][i + j - 1], max(dp[j + 1][i + j], dp[j][i + j - 1]));
            }
        }
    }

    cout << dp[0][l - 1];
    ////////////////////////////////////////////
    return 0;
}
